Closes #4378 - Refactor AbstractFetchDownloadService

R4md4c opened the pull request:
---
I've removed the dependency on `LocalBroadcastManager` and added a field in the `BrowserState` to maintain the completedDownloads ids. 

When the download completes I dispatch an action through the `BrowserStore` which has a subscription in `FetchDownloadManager` which in turn invokes `onDownloadCompleted`.

I'm not sure if the modeling of the `BrowserState` is optimal or not, I'm open for better ideas however.

EDIT: Added more explanation about what I have done. 

<!-- Text above this line will be added to the commit once "bors" merges this PR -->

### Pull Request checklist
<!-- Before submitting the PR, please address each item -->
- [x] **Quality**: This PR builds and passes detekt/ktlint checks (A pre-push hook is recommended)
- [x] **Tests**: This PR includes thorough tests or an explanation of why it does not
- [ ] **Changelog**: This PR includes [a changelog entry](https://github.com/mozilla-mobile/android-components/blob/master/docs/changelog.md) or does not need one
- [ ] **Accessibility**: The code in this PR follows [accessibility best practices](https://github.com/mozilla-mobile/shared-docs/blob/master/android/accessibility_guide.md) or does not include any user facing features

### After merge
- [ ] **Milestone**: Make sure issues closed by this pull request are added to the [milestone](https://github.com/mozilla-mobile/android-components/milestones) of the version currently in development.
- [ ] **Breaking Changes**: If this is a breaking change, please push a draft PR on [Reference Browser](https://github.com/mozilla-mobile/reference-browser) to address the breaking issues.

rocketsroger commented on components/browser/state/src/main/java/mozilla/components/browser/state/state/BrowserState.kt:
```diff
@@ -11,11 +11,14 @@ import mozilla.components.lib.state.State
 data class BrowserState(
     val tabs: List<TabSessionState> = emptyList(),
     val selectedTabId: String? = null,
 
```
nit: remove extra lines

rocketsroger replied:
nit: remove extra empty lines

rocketsroger replied:
nit: not sure why we had a empty line here.  Let's remove this line and also the extra empty line you added.

rocketsroger commented on components/browser/state/src/main/java/mozilla/components/browser/state/state/BrowserState.kt:
```diff
@@ -11,11 +11,14 @@ import mozilla.components.lib.state.State
     val selectedTabId: String? = null,
 
+    val completedDownloads: Set<Long> = emptySet(),
+
```
nit: same as above

rocketsroger commented on components/browser/state/src/main/java/mozilla/components/browser/state/state/BrowserState.kt:
```diff
@@ -11,11 +11,14 @@ import mozilla.components.lib.state.State
     val tabs: List<TabSessionState> = emptyList(),
     val selectedTabId: String? = null,
 
+    val completedDownloads: Set<Long> = emptySet(),
```
Is completed download a browser wide value or a tab specific value?  Will we ever need the tab to show that its downloading/finished downloaded?  This might require more conversation.

R4md4c replied:
The current behavior in `AndroidDownloadManager` and `AbstractFetchDownloadService` doesn't seem to contain any tab information, so I think it should go to the BrowserState, however I might be not seeing the big picture, so I'll leave the judgement to people who has more knowledge than me in the code base. 
Let me know what do you think please.

rocketsroger replied:
I agree, this one will need @pocmo to think about.

rocketsroger commented on components/browser/state/src/test/java/mozilla/components/browser/state/action/DownloadActionTest.kt:
```diff
@@ -0,0 +1,25 @@
+
+import mozilla.components.browser.state.store.BrowserStore
+import mozilla.components.support.test.ext.joinBlocking
+import org.junit.Assert
```
nit: import org.junit.Assert.assertEquals will make the test (arguable marginal) easier to read.

R4md4c commented:
Hello @pocmo , Any input please regarding @rocketsroger's comment?

pocmo commented on components/feature/downloads/src/main/java/mozilla/components/feature/downloads/manager/FetchDownloadManager.kt:
```diff
@@ -81,29 +84,31 @@ class FetchDownloadManager<T : AbstractFetchDownloadService>(
-        val downloadID = intent.getLongExtra(EXTRA_DOWNLOAD_ID, -1)
+        val downloadID = completedDownloads
+                .find { queuedDownloads.indexOfKey(it) >= 0 } ?: return
+
```
We process a download id from the set here but we never remove it from the set. I think we should add a "consume" action to remove a processed id from the store again.

pocmo commented on components/browser/state/src/main/java/mozilla/components/browser/state/action/BrowserAction.kt:
```diff
@@ -302,3 +302,13 @@ sealed class EngineAction : BrowserAction() {
+    /**
+     * Adds the [downloadId] into the set of completed downloads inside the [BrowserState].
+     */
+    data class DownloadCompletedAction(val downloadId: Long) : DownloadAction()
```
I wonder whether we could put the whole `Download` object into this action. But I assume this does not work in the case where we pass the download to Android's Download Manager (except if we keep the download object in the download manager wrapper?).

R4md4c commented:
@csadilek 
I think that's a good idea, but then IMO we would need to introduce actions like `DownloadStarted` and `DownloadFinished` to be able to maintain the queued download, right?

My only concern is that do we need to introduce that change in a "Refactor AbstractFetchDownloadService" ticket?

pocmo commented:
@csadilek I like that idea!

@R4md4c Maybe we could model this with an enum state inside the `Download` object? That would be quite nice in general if an app wants to builder UI based on the state of the Download (without needing to look at various places inside the state object).

R4md4c commented:
@pocmo @csadilek Can you please take a look on my approach of the `BrowserState` and `DownloadState`? in [1] and [2] respectively?

I've implemented that statuses taking into account that the users of the `BrowserStore` may display a UI of the current `DownloadState`s in the store.

Please note that the last commit is missing tests and other stuff, I thought about committing and validating my approach first before going further with the complete implementation.

[1] https://github.com/R4md4c/android-components/blob/f2e6fcbcecc8bbf7547672d1cc935f052eddb853/components/browser/state/src/main/java/mozilla/components/browser/state/state/BrowserState.kt#L24

[2] https://github.com/R4md4c/android-components/blob/f2e6fcbcecc8bbf7547672d1cc935f052eddb853/components/browser/state/src/main/java/mozilla/components/browser/state/state/content/DownloadState.kt#L22

pocmo commented on components/browser/state/src/main/java/mozilla/components/browser/state/reducer/DownloadStateReducer.kt:
```diff
@@ -0,0 +1,48 @@
+        is DownloadAction.DownloadFailedAction -> {
+            state.changeDownloadStatusOf(downloadState = action.download, newState = DownloadStatus.FAILED)
+        }
+    }
```
I still think we also need an action to "consume" a download here once we no longer need it in the state?

pocmo commented on components/browser/state/src/main/java/mozilla/components/browser/state/state/BrowserState.kt:
```diff
@@ -19,6 +21,7 @@ import mozilla.components.lib.state.State
 data class BrowserState(
     val tabs: List<TabSessionState> = emptyList(),
     val selectedTabId: String? = null,
+    val queuedDownloads: List<DownloadState> = emptyList(),
```
Now that we keep all state in here... can this be moved back to the individual tabs? Afaik now we do not need to have this as a global state anymore?

R4md4c commented:
@pocmo To be able to continue working on this ticket, I'd like to have some questions answered.

1. You mentioned that `queuedDownloads` needs to be per `TabSessionState`, however there is `downloadState` which is inside `ContentState`, would it still be needed if we've added queuedDownloads to the `TabSessionState`? or better put the `queuedDownloads` inside the `ContentState`?
2. Since we've introduced the DownloadAction, what is the purpose then of `ContentAction.UpdateDownloadAction` and `ContentAction.ConsumeDownloadAction` and would it make sense to keep them since DownloadActions can be used to provide the same reduction to the `ContentState`?
